doc/proposals/scorecard-user-tests.md

```diff
@@ -0,0 +1,153 @@
+                containers:
+                  - image: memcached:1.4.36-alpine
+      status:
+        scorecard_function_length:
+          nodes: 3
+    modifications:
+      - spec:
+          size: 4
+        expected:
+          resources:
+            - kind: Deployment
+              name: "example_memcached"
+              status:
+                readyReplicas: 4
+          status:
+            scorecard_function_length:
+              nodes: 4
+```
+
+This is what the golang structs would look like, including comments describing each field:
+
+```go
+// Struct containing a user defined test. User passes tests as an array using the `functional_tests` viper config
+type UserDefinedTest struct {
+    // Path to cr to be used for testing
+    CRPath string `mapstructure:"cr"`
+    // Expected resources and status
+    Expected Expected `mapstructure:"expected"`
+    // Sub-tests modifying a few fields with expected changes
+    Modifications []Modification `mapstructure:"modifications"`
+}
+
+type Expected struct {
+	// Resources expected to be created after the operator reacts to the CR
+	Resources []map[string]interface{} `mapstructure:"resources"`
+	// Expected values in CR's status after the operator reacts to the CR
+	Status map[string]interface{} `mapstructure:"status"`
+}
+
+// Modifications specifies a spec field to change in the CR with the expected results
+type Modification struct {
+    // a map of the spec fields to modify
+    Spec map[string]interface{} `mapstructure:"spec"`
+    // Expected resources and status
+    Expected Expected `mapstructure:"expected"`
+}
+```
+
+For `Status` and `Resources` fields, we can implement a bit of extra computation instead of simple string checking. For instance,
+in the memcached-operator test, we should expect that the length of the `nodes` field (which is an array) has a certain length. To implement functions like
+these, we can create some functions for these checks that are prepended by `scorecard_function_` and take an array of objects. For instance, in the above
+example, `scorecard_function_length` would check that each field listed under it matches the specified length (like `nodes: 4`). If the yaml key does not
+start with `scorecard_function_`, we do a simple match (like `status/readyReplicas: 4`).
+
+This design would allow us to replace the old "Operator actions are reflected in status" (which would be tested by the `expected/status` check) and
+"Writing into CRs has an effect" (which would be tested by the `expected/resources` check) tests.
+
+### Plugin System
+
+In order to increase the flexibility of the user defined tests and allow users to implement more complex E2E style tests for scorecard,
+the user-defined tests will be implemented via a plugin system. Users will specify the path of the script they wish to run as well
+as environment variable to set for the command. The command would then print out the result as JSON to stdout. Here is an example:
```

Comment:
Is this suggesting that you pass a yaml file into the command and it runs the things that way? 

I would like to see maybe a directory that contains the runnable (binarys/scipts) that it just runs. This will 1. make creating the image version of this easier, 2. will allow us to create system-wide tests and project-specific tests like the yaml file but using convention rather than another yaml file.

Thoughts?

Reply:
The way I was thinking to have this work would just be adding a new section to the scorecard's config file that we can read with viper. That section, which I put the example of below, would specify where the scripts are and what environment variables should be set for configuration purposes. Can you give an example of what you would like to change?

Reply:
Instead of adding a new section to the config file, I could just add 1 new flag/config option pointing to the directory where the scorecard scripts are and treat each top level file ending in `.sh` as a test. And I guess you're right about not needing the environment variables. One env var that may make sense to set for all scripts would be `$KUBECONFIG`, to make sure all scripts run in the correct cluster.

When it comes to how users should handle the executables, it depends on the user. Probably most ideal would be the script downloading the executable, similar to how we download `dep` during the CI. A user could also include it in their repo if they feel like that would be better (we do this for the `marker` executable, since it is quite tiny but would take a long time to compile during CI). As long as the script can run on a generic linux system using bash, it should be fine.

Reply:
Why can you just run all the top level files that are executable?

Reply:
@shawn-hurley Yeah, I'll update this to run top level files that are executable instead of just scripts.